lib/pkcs11h-util.c

```diff
@@ -225,30 +240,39 @@ _pkcs11h_util_unescapeString (
 	const char *s = source;
 	char *t = target;
 	size_t n = 0;
+	size_t bytes_per_esc;
+	int high_nibble_pos;
 
 	/*_PKCS11H_ASSERT (target!=NULL); Not required*/
 	_PKCS11H_ASSERT (source!=NULL);
 	_PKCS11H_ASSERT (max!=NULL);
 
 	while (*s != '\x0') {
-		if (*s == '\\') {
+		if (*s == '\\' || *s == '%') {
+			if (*s == '%') {
+				bytes_per_esc = 3;
+				high_nibble_pos = 1;
+			} else {
+				bytes_per_esc = 4;
+				high_nibble_pos = 2;
+			}
 			if (t != NULL) {
-				if (n+1 > *max) {
+				if (n+bytes_per_esc > *max) {
 					rv = CKR_ATTRIBUTE_VALUE_INVALID;
```

Comment:
@alonbl I think `n+1` should be changed to `n+4` in the current code, otherwise we could overflow the buffer.